envi/archs/i386/emu.py

```diff
@@ -2641,3 +2641,9 @@ def i_pextrw(self, op):
 
     def i_pextrd(self, op):
         self.i_pextrb(op, width=4)
+        
+    def i_vpcext(self, op):
+        # expected behavior: EBX is set to 0 if Virtual PC is detected or an exception is raised
+        # in a malware sample with an anti-vm check using this instruction, vpcext is followed by a "test ebx, ebx" and exception handling code. 
+        # to avoid a "non-defined" value in the register it is set to zero
+        self.setRegister(REG_EBX, 0)
```

Comment:
if we're trying to defeat anti-reversing techniques, shouldn't this return some other value?  it looks to me that ebx==0 if you're running in a VM... meaning, the malware would bomb.

Reply:
thank you for the feedback. I will test in the next days, if the register content of ebx is changed when running in a non-virtual environment (or the original value is kept) and will also modify the code, that the code does not end in a non-defined register value or to be caught by the following test eax, eax.

Reply:
thank you for your help, and great attitude, @richardweiss80 !  this can be a lot of fun, as i believe you've seen. :)

this is the first "non-authentic" instruction we've considered wrapping into Vivisect, so information is mostly from malware-analysis, lol.  it's not as "sure" as traditional instructions, and a little uncomfortable.  but bear with us as we work through it with you.

sooo...  it seems that the two options for emulation are:
* set ebx = 0 (which you've done)  which means that you're in a VM
* raise an envi.exc.BadOpcode exception.  since this seems to be the desired effect to defeat anti-malware, this may be the best option.  see it's use in the `ud0` instruction emulation:

```
    def i_ud0(self, op):
        raise e_exc.BadOpcode(op)
```

then, an enterprising malware researcher could subclass `vivisect.impemu.platarch.windows.Windowsi386Emulator` to handle the BadOpcode exception in , and execute the necessary exception handler, if that were interesting. :)


however, making `i_vpcext()` raise an exception would stop emulation analysis at the `vpcext` instruction.  i'm not sure that matters terribly much.
codeflow analysis (the non-emulated recursive descent analysis to find and grab initial data about the workspace) should keep going, so the following instructions should still be identified.

@rakuy0 you have any opinions here?

perhaps someday we'll implement Vivisect Analysis "modes," settable from the CLI or auto-identified, like Kernel mode, User mode, Malware mode, and emulation analysis can stem off those.  but that's not in the near future, unless someone really wants it.

Reply:
then again, that is definitely one of the reasons i wrote Vivisection, a plugin for Vivisect to do a lot of crazy extra stuff....  hmmm.

Reply:
Thank you for this detailled feedback. And you are right, this is a lot of fun.

I have taken another look at the recent samples that utilize `vpcext`; in all of them I found before the call the following instructions:
```
BB 00 00 00 00    mov ebx, 0
B8 01 00 00 00    mov eax, 1
```
Because of the following `test ebx, ebx` I think it could be a good idea to change the content of `ebx` to another value (even if it is not the correct way of working).  Also it needs more testing, which hybervisor implements `vpcext` at all.

In my opinion vivisect should set the ebx register to a value not equal to `0` explicitly. What do you think?

I will have a look at Vivisection. Thank you!

Reply:
check and see if setting `ebx` to non-zero helps or hurts.
i'm not certain the immediate best path, but i'm currently still a fan of raising a BadOpcode exception, which would allow an analysis module to catch the exception, and potentially do something valuable with it.

as for administrativa, i'm basically good with this pull request, once you change the CPU type.